### ISSUE TITLE
[Bug]: `errorbar` does not support vectorized plotting

### ISSUE THREAD

**Issue by fAndreuzzi** [matplotlib/matplotlib#23642]
### Bug summary

Since `plt.plot` is able to plot the lines of a matrix when given a 2D NumPy array, I was expecting the same from `plt.errorbar`. Instead I received an unclear error which does not really point to the problem.

### Working example with `plt.plot`
```python
import numpy as np
import matplotlib.pyplot as plt

steps = [0,1,2,3]
values = np.arange(64).reshape((4,8,2))
labels = ['val{}'.format(i) for i in range(values.shape[1])]

plt.plot(
    steps,
    values[..., 0],
    label=labels,
)
plt.legend();
```

### Code for reproduction

```python
After running code above:

plt.errorbar(
    steps,
    values[..., 0],
    yerr=values[..., 1],
    capsize=8,
    label=labels,
)
```

### Actual outcome

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [21], in <module>
----> 1 plt.errorbar(
      2     steps,
      3     values[..., 0],
      4     yerr=values[..., 1],
      5     capsize=8,
      6     label=labels,
      7 )
      8 plt.legend()

File /usr/local/lib/python3.9/site-packages/matplotlib/pyplot.py:2741, in errorbar(x, y, yerr, xerr, fmt, ecolor, elinewidth, capsize, barsabove, lolims, uplims, xlolims, xuplims, errorevery, capthick, data, **kwargs)
   2735 @_copy_docstring_and_deprecators(Axes.errorbar)
   2736 def errorbar(
   2737         x, y, yerr=None, xerr=None, fmt='', ecolor=None,
   2738         elinewidth=None, capsize=None, barsabove=False, lolims=False,
   2739         uplims=False, xlolims=False, xuplims=False, errorevery=1,
   2740         capthick=None, *, data=None, **kwargs):
-> 2741     return gca().errorbar(
   2742         x, y, yerr=yerr, xerr=xerr, fmt=fmt, ecolor=ecolor,
   2743         elinewidth=elinewidth, capsize=capsize, barsabove=barsabove,
   2744         lolims=lolims, uplims=uplims, xlolims=xlolims,
   2745         xuplims=xuplims, errorevery=errorevery, capthick=capthick,
   2746         **({"data": data} if data is not None else {}), **kwargs)

File /usr/local/lib/python3.9/site-packages/matplotlib/__init__.py:1352, in _preprocess_data.<locals>.inner(ax, data, *args, **kwargs)
   1349 @functools.wraps(func)
   1350 def inner(ax, *args, data=None, **kwargs):
   1351     if data is None:
-> 1352         return func(ax, *map(sanitize_sequence, args), **kwargs)
   1354     bound = new_sig.bind(ax, *args, **kwargs)
   1355     auto_label = (bound.arguments.get(label_namer)
   1356                   or bound.kwargs.get(label_namer))

File /usr/local/lib/python3.9/site-packages/matplotlib/axes/_axes.py:3340, in Axes.errorbar(self, x, y, yerr, xerr, fmt, ecolor, elinewidth, capsize, barsabove, lolims, uplims, xlolims, xuplims, errorevery, capthick, **kwargs)
   3334 kwargs['label'] = '_nolegend_'
   3336 # Create the main line and determine overall kwargs for child artists.
   3337 # We avoid calling self.plot() directly, or self._get_lines(), because
   3338 # that would call self._process_unit_info again, and do other indirect
   3339 # data processing.
-> 3340 (data_line, base_style), = self._get_lines._plot_args(
   3341     (x, y) if fmt == '' else (x, y, fmt), kwargs, return_kwargs=True)
   3343 # Do this after creating `data_line` to avoid modifying `base_style`.
   3344 if barsabove:

ValueError: too many values to unpack (expected 1)
```

### Expected outcome

Something similar to the output provided by `plt.plot`, with error bars:
![image](https://user-images.githubusercontent.com/8464342/185007080-71e6b495-046c-4815-9a12-b2425d60a789.png)

### Additional information

This is not a critical bug, but I feel that it's good to work on this just to keep `plt.plot` and `plt.errorbar` consistent.

I'm willing to work on this if needed :)

### Operating system

MacOS

### Matplotlib Version

3.5.3

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.9.12

### Jupyter version

6.2.0

### Installation

pip

**Comment by timhoffm** [comment]
I'm hesitant on this. First, I have to mention that we will certainly not vectorize all plotting methods, because it's not feasible (or even well-defined) for all. However, I can see the argument, that some `plot`-based methods could be vectorized as well.

In practice, vectorizing get's quite messy:
- I strongly suggest limiting the vectorization to the data. Otherwise the code get's really messy because you have to loop over everyting. E.g. all style parameters should be kept in common.
- Dimension ambiguity: Currently you can pass
  > xerr, yerr : float or array-like, shape(N,) or shape(2, N)
  for constant / symmetric per point / +/- per point.
  If you now pass two datasets and shape(2, N) errors, do you mean one shape(N,) error per dataset or do you mean one shape(2, N) error which is applied to both datasets?
  Also, if you want to pass +/- per dataset, would that be shape (2, 2, N) or shape (2, N, 2)? Accepting high-dimensional input for differenent semantic concepts is not very safe from a usability perspective. There's always the risk that users mix up the dimensions.
- Another issue we have is type-stability: We usually create one Artist per dataset and return that Artist. `plot()` was designed with the vectorization in mind and thus returns a list of `Line2D` (often with just one element). Whether type stability is really needed in that case is a separate discussion, but that's what we have. If `errorbar` is vectorized, it must also return a list in case of multiple inputs. However, for a single input it currently does not, and we cannot change that for backward-compatibility reasons. Thus, you'll end up with a structurally different return behavior compared to `plot()`.

**Comment by tacaswell** [comment]
I agree with everything that @timhoffm said.  Unfortunately given the current state of the API I do not think this is something we can reasonably add.

There is significant complexity inside of `plot` to make that broadcasting work (there is a whole callable class whose job is to handle the logic!) and the other place where we have to deal with this sort of inferring ambiguous user-intent-based-on-dimensions is in setting the color of a scatter plot.  There is a genera of bug which is "when I plot 3 points it gets color mapped, but any other number of points is the RGB value I passed!" and I strongly suspect that if we went down this route we would have an abundance of similar bugs. 

Better notions of how to inject vectorization into everything is in scope for the NASA funded work.
